### PR TITLE
Fix race conditions between hub start and stop functions.

### PR DESCRIPTION
**Pull request by slari-pivotal**
The gpupgrade pipeline https://prod.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade is failing on the CI, but not locally. The reason for this is because of a race condition between the calls to hub.Stop() and hub.Start() in the test. The reason is that hub.Stop() also closes agent connections. However, in the test, hub.Start() is called from a go routine which can sometimes start after the call to hub.Stop() has happened. Hence, the failure in hub.Stop() in the test.

We have fixed this by using a channel/flag to avoid the race condition.

Additional tests are also added.

### CODE REVIEW

File: hub/services/hub_test.go [hunk 1]
```diff
@@ -34,9 +38,66 @@ var _ = Describe("Hub", func() {
...
+			err = hub.Start()
+		}()
+		//Using Eventually ensures the test will not stall forever if this test fails.
+		Eventually(func() error { return err }).Should(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("failed to listen"))
+	})
+
+	// This is inherently testing a race. It will give false successes instead
+	// of false failures, so DO NOT ignore transient failures in this test!
+	It("will return from Start() if Stop is called concurrently", func() {
```
Comment (nadeemg) [inline hub/services/hub_test.go]: This test is trying to hit the block on lines 98-102 in hub.go? Is there anything we could do to make this more deterministic? If not, I'd vote for deleting it.

File: hub/services/hub.go [hunk 2]
```diff
@@ -26,6 +26,9 @@ import (
 
 var DialTimeout = 3 * time.Second
 
+// Returned from Hub.Start() if Hub.Stop() has already been called.
+var HubStopped = errors.New("hub is stopped")
+
```
Comment (nadeemg) [inline hub/services/hub.go]: Is this global var only being used on line 101 below?

File: hub/services/hub.go [hunk 3]
```diff
@@ -145,7 +161,9 @@ func (h *Hub) AgentConns() ([]*Connection, error) {
 	hostnames := h.clusterPair.GetHostnames()
 	for _, host := range hostnames {
 		ctx, cancelFunc := context.WithTimeout(context.Background(), DialTimeout)
-		conn, err := h.grpcDialer(ctx, host+":"+strconv.Itoa(h.conf.HubToAgentPort), grpc.WithInsecure(), grpc.WithBlock())
+		conn, err := h.grpcDialer(ctx,
+			host+":"+strconv.Itoa(h.conf.HubToAgentPort), grpc.WithInsecure(),
```
Comment (jchampio) [inline hub/services/hub.go]: If we're breaking the line anyway, can we group the configuration?
```go
conn, err := h.grpcDialer(ctx,
    host + ":" + strconv.Itoa(h.conf.HubToAgentPort),
    grpc.WithInsecure(), grpc.WithBlock())
```

File: hub/services/hub_test.go [hunk 4]
```diff
@@ -34,9 +38,50 @@ var _ = Describe("Hub", func() {
...
+		Eventually(func() error { return err }).Should(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("failed to listen"))
+	})
+
+	// This is inherently testing a race. It will give false successes instead
+	// of false failures, so DO NOT ignore transient failures in this test!
 	It("will return from Start() if Stop is called concurrently", func() {
 		hubConfig := &services.HubConfig{
-			HubToAgentPort: port,
+			CliToHubPort: cliToHubPort,
```
Comment (jchampio) [inline hub/services/hub_test.go]: Why the difference here?

File: hub/gpupgrade_hub_main.go [hunk 5]
```diff
@@ -85,7 +85,10 @@ func main() {
 				hub.MakeDaemon()
 			}
 
-			hub.Start()
+			err := hub.Start()
+			if err != nil {
+				return err
+			}
 
 			hub.Stop()
```
Comment (jchampio) [inline hub/gpupgrade_hub_main.go]: Next step (not part of this PR) is to just make 
```go
return hub.Start()
```
safe. If `Start()` returns an error, then either it failed to start entirely (so we shouldn't need to call `Stop()`) or someone else already called `Stop()` (ditto).

File: hub/services/hub_test.go [hunk 6]
```diff
@@ -34,9 +38,66 @@ var _ = Describe("Hub", func() {
...
+		}
+		hub := services.NewHub(clusterPair, grpc.DialContext, hubConfig, nil)
+
+		hub.Stop()
+		go func() {
+			err = hub.Start()
+		}()
+		//Using Eventually ensures the test will not stall forever if this test fails.
+		Eventually(func() error { return err }).Should(HaveOccurred())
```
Comment (jchampio) [inline hub/services/hub_test.go]: This should be using `services.HubStopped` for comparison, to make sure we're using the right code path.

File: hub/services/hub_test.go [hunk 7]
```diff
@@ -17,13 +19,15 @@ import (
 
 var _ = Describe("Hub", func() {
 	var (
-		agentA      *testutils.MockAgentServer
-		port        int
-		clusterPair *utils.ClusterPair
+		agentA         *testutils.MockAgentServer
+		cliToHubPort   int
+		hubToAgentPort int
```
Comment (jchampio) [inline hub/services/hub_test.go]: If we're going with prefixes, can we go with something like `hubPort` and `agentPort` instead? Save some precious columns :smile:

File: hub/services/hub.go (hunk 2 above)
Comment (jchampio) [reply hub/services/hub.go]: (see [my comment below](https://github.com/greenplum-db/gpupgrade/pull/39#discussion_r204079991))

File: hub/services/hub_test.go (hunk 1 above)
Comment (jchampio) [reply hub/services/hub_test.go]: This is a tripwire -- it helped us prove we'd fixed the problem, and if we regress it should occasionally fail to show us something is in need of investigation. Is there a specific reason you want to get rid of it?

Comment (nadeemg) [reply hub/services/hub_test.go]: Transient failures is too close to a flakey test. What is stopping is us from making this deterministic?

Comment (jchampio) [reply hub/services/hub_test.go]: > What is stopping is us from making this deterministic?
Because it's not testing a deterministic code path; the test is "make sure we don't hang when two paths are called concurrently". Adding determinism -- either by serializing the calls or forcing additional test-only serialization between the two -- will not test what was failing.

File: hub/services/hub_test.go (hunk 4 above)
Comment (kyeap-pivotal) [reply hub/services/hub_test.go]: It doesn't make any difference to the logic path, it could even be `&services.HubConfig{}`. Either way the port we pass into `CliToHubPort` is 0 (which gets us an open port), but since this is testing general race conditions pertaining to `hub.Start()` and `hub.Stop()`, setting the port that the hub listens on rather than setting the port agents listen on is much more relevant.

Comment (jchampio) [reply hub/services/hub_test.go]: Oh, cool. If nothing uses that port, I think passing in a zero-valued `HubConfig` is probably best for now.

File: hub/services/hub_test.go (hunk 1 above)
Comment (nadeemg) [reply hub/services/hub_test.go]: I've seen latches used to test scenarios like this. Have we already considered that kind of a test?

Comment (slari-pivotal) [reply hub/services/hub_test.go]: As we read it, currently there should not be any race between `Start()` and `Stop()` because if `Stop()` acquires the `h.mu.lock` before `Start()`, then `Stop()` will run to completion and `Start()` will exit out when it gets the lock.  Similarly, if `Start()` gets the lock before `Stop()`, then `Start()` would run to completion and `Stop()` would follow after that.

In case, a code path is added to `Stop()` that never releases the mutex, then a hang in `Start()` is possible, and the `Eventually(done).Should(Receive())` should catch that.

Similarly, if call to`hub.Start()` hangs for some other reason then the `Eventually()` should catch that.

Comment (nadeemg) [reply hub/services/hub_test.go]: Talked this over with @jchampio and he's going to think a little more about what we could do here. We are agreed that its not nice to have a test fail and leave it up to the dev to figure out which of the n previous green commits might have introduced a bug. We're also agreed that we dont want to jump through hoops just to get a deterministic test if that's as hard to do as it seems.

I'm fine with this going in as is.
